### ISSUE TITLE
IAM Paths not recognized in list operations

### ISSUE THREAD

**Issue by rpattcorner** [localstack/localstack#2767]
# Type of request: This is a ...

[ X] bug report
[ ] feature request

# Detailed description
Could have sworn I posted this a few days ago, but it's gone now, so here's a repeat.

AWS IAM service maintains path attributes in a range of artifacts including users, roles and policies.  Paths don't make an attribute unique (darn it!) but they do provide critical information for working on groups of artifacts with something -- usually an affiliation -- in common.

Paths are used as filters in list operations, using the `PathPrefix` optional parameter, for example

`role_list = iam.list_roles(PathPrefix='/somePathOrOther/')["Roles"]`

AWS handles path list filtering properly, but localstack ignores the PathPrefix in list operations.  From what I see, localstack returns all path entries in a list operation, regardless of filter.

## Expected behavior

iam.list_XXX operations return only entries that match the PathPrefix parameter when the parameter is present

## Actual behavior

iam.list_XXX operations return all entries regardless of PathPrefix

# Steps to reproduce

## Command used to start LocalStack

docker-compose up

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)
```
#  Show problem localstack does not honor path
import localstack_client.session

session = localstack_client.session.Session()
iam = session.client('iam')
assume_role_policy_doc = '{"Version": "2012-10-17", "Statement": [{"Effect": "Allow", "Principal": {"AWS": ["arn:aws:iam::123412341234:root"]}, "Action": "sts:AssumeRole"}]}'

# clear out roles and users for a clean test
try:
    response_role = iam.delete_role(RoleName='fooRole')
    response_user = iam.delete_user(UserName='fooUser')
except Exception as e:
    pass

# Create a unique role with a clearly defined path
response = iam.create_role(
    Path='/somePathOrOther/',
    RoleName='fooRole',
    AssumeRolePolicyDocument='{"Version": "2012-10-17", "Statement": [{"Effect": "Allow", "Principal": {"AWS": ["arn:aws:iam::123412341234:root"]}, "Action": "sts:AssumeRole"}]}',
    Description='string'
)

#  Look for the role where it *should not be found*, in the default path
#  The path format follows the aws rules of starting and ending with a slash, but a naked path gives same result
role_list = iam.list_roles(PathPrefix='/')["Roles"]
if len(role_list) != 0:
    for role in role_list:
        print ("We asked for a list of roles with a PathPrefix of '/' but found the role with name \'{}\' and path \'{}\'".
               format(role["RoleName"], role["Path"]))

#  Users have the same problem
response = iam.create_user(
    Path='/somePathOrOther/',
    UserName='fooUser',
)

#  Look for the role where it *should not be found*, in the default path
user_list = iam.list_users(PathPrefix='/')["Users"]
if len(role_list) != 0:
    for user in user_list:
        print ("We asked for a list of users with a PathPrefix of '/' but found the user with name \'{}\' and path \'{}\'".
               format(user["UserName"], role["Path"]))

#  Generally the localstack implementation of list_users and list_roles symmetrically ignores the path, e.g.
#  creating an entity with a default (/) path shows up when you ask for a more specific path

#  Clear out again
try:
    response_role = iam.delete_role(RoleName='fooRole')
    response_user = iam.delete_user(UserName='fooUser')
except Exception as e:
    pass

response = iam.create_role(
    RoleName='fooNoPathRole',   # Note, path is missing, reverts to slash
    AssumeRolePolicyDocument='{"Version": "2012-10-17", "Statement": [{"Effect": "Allow", "Principal": {"AWS": ["arn:aws:iam::123412341234:root"]}, "Action": "sts:AssumeRole"}]}',
    Description='string'
)

#  Look for the default path role where it *should not be found*, in a specific path
role_list = iam.list_roles(PathPrefix='/somePathOrOther/')["Roles"]
if len(role_list) != 0:
    for role in role_list:
        print ("We asked for a list of roles with a PathPrefix of '/somePathOrOther/' but found the role with name \'{}\' and path \'{}\'".
               format(role["RoleName"], role["Path"]))


```
With the result:
```
localstack_role_err/not_honor_path.py
We asked for a list of roles with a PathPrefix of '/' but found the role with name 'fooRole' and path '/somePathOrOther/'
We asked for a list of users with a PathPrefix of '/' but found the user with name 'fooUser' and path '/somePathOrOther/'
We asked for a list of roles with a PathPrefix of '/somePathOrOther/' but found the role with name 'fooNoPathRole' and path '/'

Process finished with exit code 0
```
...

**Comment by rpattcorner** [comment]
Thanks so much!  Looking forward to when it's merged.  I haven't had much luck getting the real latest with LATEST, designation, so it may take until there's a tagged version for me to test it.  Soon I hope - have a practical use for it!

**Comment by whummer** [comment]
Thanks for reporting @rpattcorner . This should be fixed via #2771 and  spulec/moto#3180 - can you please do another `docker pull localstack/localstack` and give it a try with latest? Please let us know if the problem persists. Thanks!
